### 학습목표
#### 다음 뉴스와 그 뉴스의 댓글 크롤링하기

In [1]:
import requests
from bs4 import BeautifulSoup

#### 뉴스 제목 크롤링

In [2]:
def get_daum_news_title(news_id):
    url = 'https://news.v.daum.net/v/{}'.format(news_id)
    resp = requests.get(url)
    
    soup = BeautifulSoup(resp.text)
    
    title_tag = soup.select_one('h3.tit_view')
    if title_tag:
        return title_tag.get_text()
    return ""

In [3]:
get_daum_news_title('20190728165812603')

'일론머스크 "테슬라에서 넷플릭스·유튜브 즐길 날 온다"'

In [6]:
get_daum_news_title('20210918151801864')

"미래 달에선 'AI 광부 로봇'들이 떼지어 자원을 캔다?"

#### 뉴스 본문 크롤링

In [7]:
def get_daum_news_content(news_id):
    url = 'https://news.v.daum.net/v/{}'.format(news_id)
    resp = requests.get(url)
    
    soup = BeautifulSoup(resp.text)
    
    content = ''
    for p in soup.select('div#harmonyContainer p'):
        content += p.get_text()
    return content

In [8]:
get_daum_news_content('20190728165812603')

'[아시아경제 이민우 기자] 일론 머스크 테슬라 최고경영자(CEO)가 자사 전기 자동차 모델에 넷플릭스와 유튜브 등 온라인동영상서비스(OTT)를 탑재할 것이라고 예고했다. 단순히 자율 주행 전기차가 단순히 주행을 위한 정보를 알려주는 것을 넘어 각종 영상 콘텐츠를 즐기는 공간으로도 확장하겠다는 전략으로 풀이된다.27일(현지시간) 더버지 등 주요 외신들에 따르면 머스크 CEO는 자신의 트위터를 통해 이 같은 계획을 밝혔다. 그는 "자동차가 정차했을 때 넷플릭스와 유튜브를 감상할 수 있는 기능이 조만간 추가될 것"이라며 "편안한 좌석과 서라운드 사운드 오디오를 통해 영화관과 같은 느낌을 받을 수 있을 것"이라고 강조했다.테슬라가 이처럼 콘텐츠 방면으로 확장하려 든 것은 이번이 처음이 아니다. 지난달 세계 최대 게임쇼 E3에서는 이미 운전자가 \'폴아웃 쉘터\'라는 게임을 할 수 있을 것이라고 발표한 바 있다. 이후에도 최근 게임업체 아타리사(社)의 자동차 경주 게임 ‘폴포지션’, 슈팅게임 ‘템페스트’, ‘미사일커맨드’ 등 고전 게임을 제공하기도 했다. 운전대로 게임을 조작하는 방식으로, 차가 주차돼 있을 경우에만 즐길 수 있다.이번 영상 콘텐츠는 주행 중에도 감상할 수 있도록 하는 방안을 고려하고 있다. 테슬라 측은 규제당국이 자율주행에 대해 완전히 승인하면 차량이 움직일 때에도 승객이 동영상을 즐길 수 있을 것이라고 설명했다.하지만 아직까지 자율주행차의 안전에 대한 우려는 완전히 걷혀지지 않은 상황이다. 지난 2017년 차량공유 서비스 우버의 자율주행 시범차량이 보행자와 충돌한 사고가 발생한 바 있다. 게다가 당시 시험 운전자는 디즈니의 동영상 스트리밍 서비스인 \'훌루\'를 이용하고 있던 것으로 밝혀졌다.이민우 기자 letzwin@asiae.co.kr<ⓒ경제를 보는 눈, 세계를 보는 창 아시아경제 무단전재 배포금지>'

In [9]:
get_daum_news_content('20210918151801864')

'[경향신문] 미래의 달 표면에서는 로봇이 떼지어 광산에서 자원을 캐는 모습을 볼 수 있을 가능성이 커졌다. 미국 과학계가 스스로 생각해 움직이는 인공지능(AI) 로봇을 달 자원 채굴에 집단적으로 투입하기 위한 연구에 나섰다.이달 중순 인셉티브 마인드 등 외신은 스스로 생각해 움직이는 자율 로봇 여러 대를 개발하려는 미국 애리조나대 과학자들에게 미국항공우주국(NASA)이 초기 연구비 50만달러(약 5억9000만원)를 지원했다고 전했다.애리조나대의 연구는 현재 미국이 주도해 한국 등 12개국이 참여하는 ‘아르테미스 계획’과 깊이 연관돼 있다. 아르테미스 계획은 2024년까지 사람을 달에 다시 보내고, 2028년부터는 사람이 항상 머무는 상주 기지를 운영하는 것이 목표다. 상주 기지를 지으려는 가장 큰 이유는 달에 오랜 기간을 두고 캐낼 가치 있는 자원이 많기 때문이다. 달 표면에는 핵융합 발전의 원료로 쓸 수 있는 물질인 ‘헬륨3’가 다량 존재한다. 헬륨3 1g은 무려 석탄 40t과 비슷한 에너지를 내는데, 달에 묻혀 있을 것으로 추정되는 헬륨3는 무려 100만t에 이른다.달에는 ‘희토류’도 풍부하다. 희토류는 네오디뮴 등 17개 원소를 지칭한다. 태양광 발전장비나 고강도 합금, 첨단 전자기기 등을 만들려면 반드시 필요한 산업계의 ‘비타민’ 같은 물질인데, 지구에선 중국이 세계 공급망을 장악하고 있다. 이외에도 티타늄이나 백금 등 귀한 금속이 달에는 많다. 달에서 광산을 개척하는 일은 기존의 자원 공급 체계를 완전히 재편할 ‘사건’인 것이다. 자원을 캐기 위해 상주 기지를 만들려면 달 표면에 건설 자재를 갖다 놔야 한다. 하지만 고도 수백㎞에 불과한 지구 저궤도에 물체를 올리는 데에도 1㎏당 로켓 발사 비용이 보통 수천만원씩 들어간다. 달까지 건설 자재를 운송하는 데에는 더 많은 비용을 써야 한다. 달 표면에서 콘크리트와 유사한 성질을 지닌 물질을 캐낸 뒤 기지 건설에 바로 쓰는 게 효율적이라는 뜻이다.애리조나대 연구진은 달에서 자원을 캐내는 것은 물론 건설 자

#### 뉴스 댓글 크롤링

댓글은 동적으로 추가로 계속 가져와야 하는 형태.

![api5](https://user-images.githubusercontent.com/33708512/133919692-f03568e2-a8fd-48eb-a379-45364f788952.png)

In [25]:
url = 'https://comment.daum.net/apis/v1/posts/@20190728165812603/comments?parentId=0&offset=43&limit=10&sort=RECOMMEND&isInitial=false'

requests.get(url)

<Response [401]>

401은 권한이 없다는 에러이고, 개발자 도구에서 확인하는 것처럼 아래와 같이 헤더를 모두 명시해서 실행하면 에러 없이 실행가능 함.

In [22]:
url = 'https://comment.daum.net/apis/v1/posts/133493400/comments?parentId=0&offset=13&limit=10&sort=POPULAR&isInitial=false&hasNext=true&randomSeed=1632032287'
headers = {
        'Authorization': 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJmb3J1bV9rZXkiOiJuZXdzIiwiZ3JhbnRfdHlwZSI6ImFsZXhfY3JlZGVudGlhbHMiLCJzY29wZSI6W10sImV4cCI6MTYzMjA3NTI4MSwiYXV0aG9yaXRpZXMiOlsiUk9MRV9DTElFTlQiXSwianRpIjoiNDg2OTE5MGMtYTAzMC00M2IxLTg4ZmUtZWRjZDQyNDFiY2Y5IiwiZm9ydW1faWQiOi05OSwiY2xpZW50X2lkIjoiMjZCWEF2S255NVdGNVowOWxyNWs3N1k4In0.O0u7_42aSetx7qi37uurCkn9JA9zOueXMgd5JASyLog',
        'Origin': 'https://news.v.daum.net',
        'Referer': 'https://news.v.daum.net/v/20190728165812603',
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/93.0.4577.63 Safari/537.36'
}

requests.get(url, headers=headers)

<Response [200]>

In [23]:
resp = requests.get(url, headers=headers)

resp.json()

[{'id': 393704000,
  'userId': 11704882,
  'postId': 133493400,
  'forumId': -99,
  'parentId': 393676277,
  'type': 'COMMENT',
  'status': 'S',
  'flags': 256,
  'rating': 0,
  'content': '스마트 시티에선 가능 할거 같음 근데 중국이 추진하는 스마트 시티는 개인신상 다털리니 안됨',
  'createdAt': '2019-07-28T19:33:02+0900',
  'updatedAt': '2019-07-28T19:33:02+0900',
  'childCount': 0,
  'likeCount': 0,
  'dislikeCount': 0,
  'recommendCount': 0,
  'screenedByKeeper': False,
  'user': {'id': 11704882,
   'status': 'S',
   'type': 'USER',
   'flags': 0,
   'icon': 'https://k.kakaocdn.net/dn/kvxJk/btrc4ESc7r4/qkeNYMJzMe1eWSXdWlkCf0/img_110x110.jpg',
   'url': 'https://story.kakao.com/_jFHkT',
   'username': 'KAKAO:218344087',
   'roles': 'ROLE_USER,ROLE_KAKAO,ROLE_IDENTIFIED',
   'providerId': 'KAKAO',
   'providerUserId': '11876290',
   'displayName': '경호',
   'description': '',
   'commentCount': 556},
  'parent': {'id': 393676277,
   'userId': -557848722,
   'postId': 133493400,
   'forumId': -99,
   'parentId': 0,
   'type'

In [27]:
def get_daum_news_comments(news_id):
    headers = {
        'Authorization': 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJmb3J1bV9rZXkiOiJuZXdzIiwiZ3JhbnRfdHlwZSI6ImFsZXhfY3JlZGVudGlhbHMiLCJzY29wZSI6W10sImV4cCI6MTYzMjA3NTI4MSwiYXV0aG9yaXRpZXMiOlsiUk9MRV9DTElFTlQiXSwianRpIjoiNDg2OTE5MGMtYTAzMC00M2IxLTg4ZmUtZWRjZDQyNDFiY2Y5IiwiZm9ydW1faWQiOi05OSwiY2xpZW50X2lkIjoiMjZCWEF2S255NVdGNVowOWxyNWs3N1k4In0.O0u7_42aSetx7qi37uurCkn9JA9zOueXMgd5JASyLog',
        'Origin': 'https://news.v.daum.net',
        'Referer': 'https://news.v.daum.net/v/20190728165812603',
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/93.0.4577.63 Safari/537.36'
    }
    
    url_template = 'https://comment.daum.net/apis/v1/posts/{}/comments?parentId=0&offset={}&limit=10&sort=POPULAR&isInitial=false&hasNext=true&randomSeed=1632032287'
    offset = 0
    comments = []
    while True:
        url = url_template.format(news_id, offset)
        resp = requests.get(url, headers=headers)
        data = resp.json()
        if not data:
            break
            
        comments.extend(data)
        offset += 10
        
    return comments


In [29]:
len(get_daum_news_comments('133493400'))

15

In [31]:
len(get_daum_news_comments('160528151'))

6